sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/CloudEventRequestContent.cs

```diff
@@ -0,0 +1,93 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Core.Pipeline;
+using Azure.Core.Serialization;
+
+namespace Azure.Messaging.EventGrid
+{
+    internal class CloudEventRequestContent : RequestContent
+    {
+        private IEnumerable<CloudEvent> _cloudEvents;
+        private static readonly JsonObjectSerializer s_jsonSerializer = new JsonObjectSerializer();
+        private const string TraceParentHeaderName = "traceparent";
+        private const string TraceStateHeaderName = "tracestate";
+        private byte[] _data;
+
+        public CloudEventRequestContent(IEnumerable<CloudEvent> cloudEvents)
+        {
+            _cloudEvents = cloudEvents;
+        }
+
+        public override void Dispose()
+        {
+        }
+
+        public override bool TryComputeLength(out long length)
+        {
+            EnsureSerializedAsync(false, CancellationToken.None).EnsureCompleted();
+            length = _data.Length;
+            return true;
+        }
+
+        public override void WriteTo(Stream stream, CancellationToken cancellationToken)
+        {
+            EnsureSerializedAsync(false, cancellationToken).EnsureCompleted();
+            stream.Write(_data, 0, _data.Length);
+        }
+
+        public override async Task WriteToAsync(Stream stream, CancellationToken cancellationToken)
+        {
+            await EnsureSerializedAsync(true, cancellationToken).ConfigureAwait(false);
+            await stream.WriteAsync(_data, 0, _data.Length, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task EnsureSerializedAsync(bool async, CancellationToken cancellationToken)
+        {
+            if (_data != null)
+            {
+                return;
+            }
+
+            string activityId = null;
+            string traceState = null;
+            Activity currentActivity = Activity.Current;
```

Comment:
We might need to invalidate the cache when the activity changes.

Reply:
As long as the activity that is used here is the same as the RequestActivityPolicy activity, we want to use that since that will be what is stamped onto the Request. If the activity changes later, and this is serialized again for some reason I don't think we would want to change anything.

Reply:
The activity would change during retries but you won't re-serialize.

Reply:
Actually, the request header doesn't change - https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/src/Pipeline/Internal/RequestActivityPolicy.cs#L117

Though this is probably a bug as I would expect it to be updated.

Reply:
Can you file that bug please?

Reply:
https://github.com/Azure/azure-sdk-for-net/issues/18992